### ISSUE TITLE
Correctly handle redirect rules

### ISSUE THREAD

**Issue by dhowe** [dhowe/AdNauseam#1771]
Branch [ublock-v1.33.0]

- [x] implement proper handling of redirect rules
- [x] add testing of redirect rules to release-testing

from #1764:
> The function `toURL` from the `src/js/redirect-engine.js` doesn't exist anymore in uBlock, generating an **exceptionError** in **Adn**. 
> It is used inside the `isBlockableRequest` function in the `src/js/adn/core.js` file:
> ```
>    // always allow redirect blocks from lists (?)
>     if (µb.redirectEngine.toURL(context)) {
>       logNetBlock('*Redirect*', context.docDomain + ' => ' + context.url, context);
>       return true;
>     }
> ```
> The new equivalent function in the `src/js/redirect-engine.js` is `tokenToURL`, but we don't have the necessary parameters in the above code to run it.
> 
> The commit in which this change occurred in uBlock: https://github.com/gorhill/uBlock/commit/157cef6034a8ec926c1e59c7e77f0a1fcbef473c

See general info on redirect rules [below](https://github.com/dhowe/AdNauseam/issues/1771#issuecomment-800369823)

See list of included redirect resources here: https://github.com/gorhill/uBlock/wiki/Resources-Library#url-specific-sanitized-redirect-resources-surrogates

See definition of RedirectEngine.tokenToURL()  here: https://github.com/gorhill/uBlock/blob/40c145d76a238255e3388d743ab2ce2194ee73b4/src/js/redirect-engine.js#L306

Token returned from: https://github.com/gorhill/uBlock/blob/1024f60fc9db37b7fd85e85c14400c0f8b36b1e1/src/js/static-net-filtering.js#L4237 (start debugging here?)

**Comment by dhowe** [comment]
## Info on Redirect Rules
---------------


### redirect

The redirect option means "block and redirect", and really causes two filters to be created internally, a block filter and a redirect directive (redirect-rule).

*A redirect directive causes a blocked network request to be redirected to a local neutered version of the resource. The neutered resource must be referenced using a resource token. The resources are defined in Resources Library. Filters with unresolvable resource token at runtime, will be discarded.*

`redirect=` filters can be used with any other static filter options, can be excepted using @@, can be badfilter-ed and their priority can be increased with the `important` option.

Since more than one redirect directives could be found to apply to a single network request, the concept of redirect priority is introduced.

By default, redirect directives have an implicit priority of 0. Filter authors can declare an explicit priority by appending `:[integer]` (negative values are also supported) to the token of the `redirect=` option, for example:

    ||example.com/*.js$1p,script,redirect=noopjs:100

The priority dictates which redirect token out of many will be ultimately used. Cases of multiple `redirect=` directives applying to a single blocked network request are expected to be rather unlikely. All directives will be reported in the logger, and the effective one will be reported as the last one before redirection entry. Explicit redirect priority should be used if and only if there is a case of redirect ambiguity to solve.

To disable a redirection, you can use an exception filter for the redirect directive specifically (example for the filter above):

    @@||example.com/*.js$1p,script,redirect-rule=noopjs

The filter above does not affect blocking filters, just matching redirect directives. You can broadly disable all redirect directives as follow:

    @@||example.com/*.js$1p,script,redirect-rule

---------------------

[Note: not sure what the below means for ADN if built on v1.33.0]

#### Before 1.32.0

Starting with 1.31.0 redirect= option is not longer afflicted by static network filtering syntax quirks listed below.

    Resource type must be specified.
    Special, reserved token none must be used to disable specific redirect filters.
    Negated domains in domain= option are not supported because of syntax ambiguity - #310.
    Redirections applied to all destinations (starting with *) cannot be narrowed by first-party or ~third-party option #3590
    Disable redirection by specifying none as the redirect. (Broken in 1.31.0, fixed in 1.31.3b4)
    Filters with unresolvable resource token at runtime will cause redirection to fail. (Changed in 1.31.1b8)

Available since 1.4.0.

**Comment by dhowe** [comment]
So the question here is whether to allow blocks that are part of redirect rules?

Previously, we did allow blocks for ALL redirect rules. However this requires knowing (here) whether we have such a case.
This is question 1.

Ideally if the answer to question 1 was true, then we would ask question 2, whether the block might contain ads or just other malfeasance, because we will (under the previous policy) NOT be able to collect any ads on within resources blocked as part of a redirect rule. 

Without the answer to question #1, we can only not-block redirects, which will potentially result in broken pages on ADN that render correctly on uB, and this is of course to be avoided.

Thus what we need to be able to do here is to distinguish redirects from other types of rules without blocking or allowing them.

**Comment by mneunomne** [comment]
@dhowe I have one test case which tells that Adn current 1.33.0 merge is handling the redirects correctly. This does not answer our question but at least now I understand how to identify such cases. Is the here workflow correct?

- __url:__ https://g1.globo.com/
- __redirected file:__ `googletagservices_gpt.js`
- __result:__ File successfully redirected, original file request with status `307` (redirect), loading instead local uBlock file. Nothing seems to break in AdNauseam. 

### uBlock test

![Screen Shot 2021-03-22 at 20 14 19](https://user-images.githubusercontent.com/4967860/112045517-3ac92180-8b4b-11eb-86c7-2f3bbb7f7537.png)

### AdNauseam test

![Screen Shot 2021-03-22 at 20 03 17](https://user-images.githubusercontent.com/4967860/112045974-bc20b400-8b4b-11eb-85c4-115ee8af0adf.png)

**Comment by mneunomne** [comment]
> So the question here is whether to allow blocks that are part of redirect or redirect-rule filters.
> 
> Previously, we _did_ allow blocks for ALL redirect filters. However this requires knowing in `core.isBlockableResource()` whether we have such a case. This is Question 1.
> 
> Ideally if the answer to Question 1 is true, then we would ask Question 2: whether the block might contain ads or just other crap, because we will NOT be able to collect ads within resources blocked as part of a redirect rule. If such rules are used to block common ads (e.g., google-ads, for which there are redirect rules) then we may find large numbers of common ads node being registeres or collected by ADN. In such a case we will _need_ to readdress Question 2.
> 
> But the first priority is Question 1, b/c without an accurate answer we can only not-block redirects, which will potentially result in broken pages on ADN that render correctly on uB, and this is of course to be strictly avoided.
> 
> Thus (Q1) what we need to be able to do here is to distinguish the 2 kinds of redirect filters from other types of rules without actually blocking them. I've implemented a fix ([89f12eb](https://github.com/dhowe/AdNauseam/commit/89f12eb6ce3544c847012060ad71d94b3d280dc6)) that _should_ do this and thus restore prior behavior. But we need to add good test cases with redirect rules (which should then fire on relevant pages, and should show up in both console, with `debugging` enabled, _and_ in the logger) to know for sure.

Regarding this I am still trying to understand it correctly. 

- If I understood correctly, with the current behaviour we are potentially missing some ads for collection because all these scripts are being redirected?
- How can we know which types of redirected scripts would be generating ads and the ones who don't ?

**Comment by dhowe** [comment]
> If I understood correctly, with the current behaviour we are potentially missing some ads for collection because all these scripts are being redirected?

Yes, if one of these scripts contains an ad, then b/c the script is not even called, the ad will of course never make it to ADN

> How can we know which types of redirected scripts would be generating ads and the ones who don't ?

This answer to this question (which applies for ALL blocking rules, not just redirects) is through a combination of testing and paying attention to what different lists do. Thus far we maintain a set of non-ad lists where we allow blocking (unless disabled by the user via the 'block malware' option). On all other lists, we disallow blocks (these are called 'adn-allow' rules and are documented in the FAQ and dev-docs)

Ideally we would do something similar for redirects

**Comment by mneunomne** [comment]
> Actually, please add a new ticket to review (post-release) the lists in core::enabledBlockLists

Yes, I was actually going ask about this after taking a look on the core.js file. But how exactly should they be reviewed?

**Comment by dhowe** [comment]
Pls add this question in the new ticket

**Comment by mneunomne** [comment]
Okay, now I finally found an example which has BOTH types of rules:
https://www.businessinsider.com/electric-car-sales-decline-2016-10?r=DE&IR=T

I found it going though redirect rules in the uBlock filter list, and checking the associated issues in github. It seems like the best way to find them:
https://github.com/uBlockOrigin/uAssets/issues/157  

The merge1.33.0 version AdNauseam Logger detects it:
![Screen Shot 2021-03-25 at 19 37 05](https://user-images.githubusercontent.com/4967860/112526300-13b85d00-8da2-11eb-85c0-f721845210a5.png)
(the name of the file lol)
![Screen Shot 2021-03-25 at 19 37 12](https://user-images.githubusercontent.com/4967860/112526290-11560300-8da2-11eb-8c4b-150b3633571b.png)
 
But it does not appear in the console, so it is not passing through our `isRedirectRule` condition:
```
if (isRedirectRule(context)) {
      logNetBlock('*Redirect*', context.docDomain + ' => ' + context.url, context);
      return true;
    }
``` 
I also checked in the requests and no request is made for such a file, so I imagine it was completely blocked? Differently from the other example I put above where the redirect was appearing in the page requests.

**Comment by mneunomne** [comment]
Another example, where both cases happen and both are being correctly logged from Adn:
https://m.bild.de/###wt_ref=https%3A%2F%2Fwww.bild.de%2F&wt_t=1616698301531

Visually checked and both in uBlock and Adn the page seems to behave the same.

### Filters:

redirect:
```
||tagger.opecloud.com^$xhr,redirect=noop.txt,domain=bild.de
```
redirect-rule:
```
||tags.tiqcdn.com/utag/*/prod/utag.js$script,redirect-rule=noopjs
```

### Adn Logger

![Screen Shot 2021-03-25 at 19 55 46](https://user-images.githubusercontent.com/4967860/112528521-888c9680-8da4-11eb-813f-69ebfd08ef8d.png)
![Screen Shot 2021-03-25 at 19 55 43](https://user-images.githubusercontent.com/4967860/112528525-89bdc380-8da4-11eb-8757-bb71e970c57f.png)

### Adn Console

![Screen Shot 2021-03-25 at 19 54 45](https://user-images.githubusercontent.com/4967860/112528571-9b06d000-8da4-11eb-9dd6-86d6a7aa63f2.png)
![Screen Shot 2021-03-25 at 19 54 39](https://user-images.githubusercontent.com/4967860/112528575-9c37fd00-8da4-11eb-8569-544b45fcaede.png)

**Comment by dhowe** [comment]
note that console.log is not synchronous, so if an object's property has changed between when it is qeued, and when it is actually output to the console, then the new value will show (which is what I imagine is happening above with the context object) -- thus one needs to serialize it in some way to see the real value (JSON.stringify is one option)

regarding the 2 types of redirect rules, can you summarize your findings? both fire but only one shows in the the debug output... or ... ?

**Comment by mneunomne** [comment]
@dhowe now reviewing actually both of them are being fired and appearing on debug output. 

I was missing the "fuckadblock.js-3.2.0" the debug console on first example because it is actually a filter detects multiple "block adblock" libraries such as `FuckAdBlock`, `BlockAdBlock`, `SniffAdBlock`, `fuckAdBlock`, `blockAdBlock`, `sniffAdBlock` as documented [here](https://github.com/gorhill/uBlock/wiki/Resources-Library#fuckadblockjs-320-). So the actual name of the file i was not finding in the Debug Output was [advertising.js](https://www.npttech.com/advertising.js) which (as you'll see) contains `blockAdBlock` property in it, which fires the redirect apparently.

![Screen Shot 2021-03-26 at 19 20 20](https://user-images.githubusercontent.com/4967860/112676108-547aa980-8e68-11eb-813a-77211c237e9e.png)

My conclusion being: both types of redirects are working and being picked up by the redirect condition. I believe we can use both examples here documented as test cases for future releases that deal with redirect somehow.

**Comment by dhowe** [comment]
Great. Then this can be closed (please do add the above examples to the release testing checklist)

**Comment by dhowe** [comment]
reopen

**Comment by dhowe** [comment]
@mneunomne pls confirm the fix and close

**Comment by mneunomne** [comment]
Tested and the behaviour of the redirect is the same as in uBlock origin. 

But still the console messages from Adn are really confusing. I can identify in [this page](bild.de/###wt_ref=https%3A%2F%2Fwww.bild.de%2F&wt_t=1616698301531) the redirect properly in the logger and find the corresponding message on the Adn console, but it is amongst hundreds of other `(*Redirect*)` messages that are actually just blocks. For example:
```
[BLOCK] (*Redirect*) bild.de => https://confiant-integrations.global.ssl.fastly.net/7GPxwkHfdSwn-OphtKZwIPivffk/axel/config.js ||confiant-integrations.global.ssl.fastly.net^ https://confiant-integrations.global.ssl.fastly.net/7GPxwkHfdSwn-OphtKZwIPivffk/axel/config.js 
```
or
```
[BLOCK] (*Redirect*) bild.de => https://www.asadcdn.com/adlib/libmodules/ovkvmf/isVisible.85e9f180.js ||asadcdn.com^$3p https://www.asadcdn.com/adlib/libmodules/ovkvmf/isVisible.85e9f180.js
``` 
(together with hundreds of such logs of block rules)

Why all of this is being logged as a redirect by Adn?

**Comment by mneunomne** [comment]
## A new test case trying to reproduce "redirect=" filter

- Open the following page that we have been using for testing: https://www.bild.de/unterhaltung/royals/leute/prinz-philip-99-ist-tot-queen-elizabeth-ii-trauert-ihr-mann-starb-auf-windsor-ca-76006958.bild.html

- Disable the following filters in the "Filter List" so we can manually add our own redirect rule: `["Easy List", "uBlock Filters", "EasyPrivacy"]`

- After doing that and reloading the page you'll see that **bild.de** detects the add blocker and blocks the page after a few seconds.

- At the logger you won't be seeing the "tagger" being blocked any longer as it is normally is from this filter `||tagger.opecloud.com^`

- Now add the following filter in the "My Filters": 
```
||tagger.opecloud.com^$xhr,redirect=noop.txt,domain=bild.de
```
- Reload the page and you'll see that Adn is actually not redirecting the request, and it is allowing **Adn-allowing** it:
![Screen Shot 2021-04-09 at 14 58 21](https://user-images.githubusercontent.com/4967860/114185707-4fc4f380-9946-11eb-94fd-5d6113fe60f5.png)

- The same behavior on uBlock generates the redirect as expected:
![Screen Shot 2021-04-09 at 14 58 50](https://user-images.githubusercontent.com/4967860/114186376-0aed8c80-9947-11eb-9658-72c96cb2ebfe.png)

Perhaps that is expected behavior since Adn deals with block in a particular manner. But at least this is a case of `redirect=` filter being properly identified.

**Comment by mneunomne** [comment]
@dhowe and btw the console seems to be much better now. Only the redirects are being logged.

**Comment by cqx931** [comment]
It's interesting that in the case @mneunomne mentioned above, the `||tagger.opecloud.com^$xhr,redirect=noop.txt,domain=bild.de` filter is interpreted as:
`||tagger.opecloud.com^$xmlhttprequest,domain=bild.de`, which seems like the redirect part has been ignored?

Meanwhile, I can't reproduce the `redirect=` case above in ublock origin somehow.

The other case `redirect-rule=` seems to work and the block rule and the redirect rule got parsed differently in this case:
block rule -> EasyPrivacy
redirect-rule -> ublock unbreak
<img width="1212" alt="Screenshot 2021-04-17 at 10 05 21 AM" src="https://user-images.githubusercontent.com/3056154/115106323-6d0c4a00-9f64-11eb-92ec-f46153eb8323.png">

Question: is there any concrete case where we can identify a redirect rule stopping an visual ad being rendered? So far all the examples seem to be either privacy or unbreak oriented.

**Comment by cqx931** [comment]
Another example

1. Go to https://gamer.nl/artikelen/interview/new-pokemon-snap-heeft-het-concept-van-de-nintendo-64-pokemon-snap-als-basis-behouden/ using latest ublock origin
2. In the logger you can see this:
<img width="1103" alt="Screenshot 2021-04-17 at 10 27 08 AM" src="https://user-images.githubusercontent.com/3056154/115106861-7ba83080-9f67-11eb-8b92-e2d56fbf1283.png">
This is triggerd by the folllowing rule in ublock filters 
`||static.chartbeat.com/js/chartbeat.js$script,redirect=static.chartbeat.com/chartbeat.js,domain=gamer.nl`

Notice how this rule is interpreted as two rules in the logger and there is never `redirect=` in the logger.
`||static.chartbeat.com/js/chartbeat.js$script,domain=gamer.nl`
`||static.chartbeat.com/js/chartbeat.js$script,redirect-rule=static.chartbeat.com/chartbeat.js,domain=gamer.nl`

**Comment by cqx931** [comment]
@dhowe Is there still anything else need to be done for redirect rules?

**Comment by dhowe** [comment]
as long as the redirect does not fire if we adn-allow the block, then this seems ok
thanks for your help, @cqx931 !
